kernel-rs/src/file.rs

```diff
@@ -176,53 +165,58 @@ impl File {
         if !self.writable {
             return -1;
         }
-        if self.typ == Filetype::PIPE {
-            self.pipe.write(addr, n)
-        } else if self.typ == Filetype::DEVICE {
-            if (self.major) < 0
-                || self.major as usize >= NDEV
-                || DEVSW[self.major as usize].write.is_none()
-            {
-                return -1;
-            }
-            DEVSW[self.major as usize]
-                .write
-                .expect("non-null function pointer")(1, addr, n)
-        } else if self.typ == Filetype::INODE {
-            // write a few blocks at a time to avoid exceeding
-            // the maximum log transaction size, including
-            // i-node, indirect block, allocation blocks,
-            // and 2 blocks of slop for non-aligned writes.
-            // this really belongs lower down, since write()
-            // might be writing a device like the console.
-            let max = (MAXOPBLOCKS - 1 - 1 - 2) / 2 * BSIZE;
-            let mut i: i32 = 0;
-            while i < n {
-                // TODO : rename `n1`
-                let n1 = cmp::min(n - i, max as i32);
-                begin_op();
-                (*self.ip).lock();
-                let r = (*self.ip).write(1, addr.wrapping_add(i as usize), self.off, n1 as u32);
-                if r > 0 {
-                    self.off = (self.off).wrapping_add(r as u32)
-                }
-                (*self.ip).unlock();
-                end_op();
-                if r < 0 {
-                    break;
+
+        // Use &mut self.typ because write() "changes" FileType::INODE.off during holding ip's lock.
+        match &mut self.typ {
+            Filetype::PIPE { pipe } => pipe.write(addr, n),
+            Filetype::INODE { ip, off } => {
+                // write a few blocks at a time to avoid exceeding
+                // the maximum log transaction size, including
+                // i-node, indirect block, allocation blocks,
+                // and 2 blocks of slop for non-aligned writes.
+                // this really belongs lower down, since write()
+                // might be writing a device like the console.
+                let max = (MAXOPBLOCKS - 1 - 1 - 2) / 2 * BSIZE;
+                let mut i: i32 = 0;
+                while i < n {
```

Comment:
무조건 `bytes_to_write` 만큼 읽는 것이 확실하므로,
```suggestion
                for i in (0..n).step_by(max) {
```
이런 식으로 functional 하게 작성할 수 있을 듯 해요. 그리고 `i`에 좀 더 설명적인 이름을 붙여주시기를 부탁드립니다.

Reply:
https://github.com/kaist-cp/rv6/pull/211/commits/192e38cb7911c76b3ac036b92e91b4016e075769